File: iOS/SampleApp/OTKAnalytics.m
//
//  OTKAnalytics.m
//  SampleApp
//
//  Created by mserrano on 09/03/16.
//  Copyright © 2016 AgilityFeat. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import <sys/utsname.h>
#import "OTKAnalytics.h"


@implementation OTKAnalyticsData

-(instancetype)initWithSessionId:(NSString*) sessionId connectionId:(NSString*) connectionId partnerId:(NSInteger) partnerId clientVersion:(NSString*) clientVersion {
    
    self = [super init];
    
    if ( sessionId == nil || [ sessionId length ]==0 ) {
        NSLog (@"The sessionId field cannot be null in the log entry");
        return nil;
    }
    
    if ( connectionId == nil || [ connectionId length ]==0 ) {
        NSLog (@"The connectionId field cannot be null in the log entry");
        return nil;
    }
    
    if ( partnerId == 0 ) {
        NSLog (@"The partnerId field cannot be null in the log entry");
        return nil;
    }
    
    if ( clientVersion == nil || [ clientVersion length ]==0 ) {
        NSLog (@"The clientVersion field cannot be null in the log entry");
        return nil;
    }
    
    self.sessionId = sessionId;
    self.connectionId = connectionId;
    self.partnerId = partnerId;
    self.clientVersion = clientVersion;
    
    return self;
}

@end

@implementation OTKAnalytics

NSString *const kLoggingUrl = @"https://hlg.tokbox.com/prod/logging/ClientEvent";

OTKAnalyticsData *_data;

-(instancetype)initWithData:(OTKAnalyticsData *)data {
    
    if ( data == nil ) {
        NSLog (@"The data cannot be null in the log entry");
        return nil;
    }
    
    _data = data;
    
    return self;
}

- (void) checkData {
    
    NSException *e = nil;
    
    if ( _data.logVersion == nil || [ _data.logVersion length ]==0 ) {
        _data.logVersion = @"2";
    }
    
    if ( _data.guid == nil || [ _data.guid length ]==0 ) {
        _data.guid = [[NSUUID UUID] UUIDString];
    }
    
    if ( _data.deviceModel == nil || [ _data.deviceModel length ]==0 ) {
        struct utsname systemInfo;
        uname(&systemInfo);
        _data.deviceModel = [NSString stringWithCString:systemInfo.machine
                                               encoding:NSUTF8StringEncoding];
    }
    
    if ( _data.client == nil || [ _data.client length ]==0 ) {
        _data.client = @"native";
    }
    
    if ( _data.clientSystemTime == 0 ) {
        NSTimeInterval timeInMiliseconds = [[NSDate date] timeIntervalSince1970];
        NSInteger time = round(timeInMiliseconds);
        _data.clientSystemTime = time;
    }
    
    if ( _data.systemName == nil || [ _data.systemName length ]==0 ) {
        _data.systemName = @"iOS OS";
    }
    
    if ( _data.systemVersion == nil || [ _data.systemVersion length ]==0 ) {
        _data.systemVersion = [[UIDevice currentDevice] systemVersion];
    }
    
}

-(void)logEventAction:(NSString *)action variation:(NSString *) variation {
    
    _data.action = action;
    _data.variation = variation;
    
    [self checkData];
    
    NSDictionary *dictionary = @{
                                 @"sessionId" : _data.sessionId,
                                 @"connectionId" : _data.connectionId,
                                 @"partnerId" : [NSNumber numberWithInteger:_data.partnerId],
                                 @"client" : _data.client,
                                 @"logVersion" : _data.logVersion,
                                 @"clientVersion" : _data.clientVersion,
                                 @"clientSystemTime" : [NSNumber numberWithInteger:_data.clientSystemTime],
                                 @"guid" : _data.guid,
                                 @"deviceModel": _data.deviceModel,
                                 @"systemName": _data.systemName,
                                 @"systemVersion": _data.systemVersion,
                                 @"action": _data.action,
                                 @"variation": _data.variation
                                 };
    
    [self sendData: dictionary];
    
}

-(void) sendData: (NSDictionary*) data {
    NSError *error;
    
    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:configuration delegate:nil delegateQueue:nil];
    NSURL *url = [NSURL URLWithString: kLoggingUrl];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url
                                                           cachePolicy:NSURLRequestUseProtocolCachePolicy
                                                       timeoutInterval:60.0];
    
    [request addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request setHTTPMethod:@"POST"];
    
    NSData *postData = [NSJSONSerialization dataWithJSONObject:data options:0 error:&error];
    [request setHTTPBody:postData];
    
    NSURLSessionDataTask *postDataTask = [session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
        NSLog(@"response status code: %ld", (long)[httpResponse statusCode]);
    }];
    
    [postDataTask resume];
    
}

@end

File: av-sample-app/js/public/js/components/otkanalytics.js
var OTKAnalytics = (function() {

    var OTKAnalyticsComponent = function(data) {
        this.url = 'https://hlg.tokbox.com/prod/logging/ClientEvent';

        this.analyticsData = data;
    }

    var _checkData = function() {
        var self = this;
        
        if ( self.analyticsData.sessionId === null || self.analyticsData.sessionId.length === 0){
            console.log ("Error. The sessionId field cannot be null in the log entry");
            throw("The sessionId field cannot be null in the log entry");
        }
        if ( self.analyticsData.connectionId == null || self.analyticsData.connectionId.length == 0){
            console.log ("Error. The connectionId field cannot be null in the log entry");
            throw("The connectionId field cannot be null in the log entry");
        }
        if ( self.analyticsData.partnerId == 0){
            console.log ("Error. The partnerId field cannot be null in the log entry");
            throw("The partnerId field cannot be null in the log entry");
        }
        if ( self.analyticsData.clientVersion == null || self.analyticsData.clientVersion.length == 0){
            console.log ("Error. The clientVersion field cannot be null in the log entry");
            throw("The clientVersion field cannot be null in the log entry");
        }
        if ( self.analyticsData.logVersion == null || self.analyticsData.logVersion.length == 0){
            self.analyticsData.logVersion  = "2";
        }
        if ( self.analyticsData.guid == null || self.analyticsData.guid.length == 0){
            self.analyticsData.guid  = _generateUuid();
        }
        if ( self.analyticsData.clientSystemTime == 0){
            self.analyticsData.clientSystemTime  = new Date().getTime();;
        }    
    };

    var _generateUuid = function() {

        // http://www.ietf.org/rfc/rfc4122.txt
        var s = [];
        var hexDigits = "0123456789abcdef";
        for (var i = 0; i < 36; i++) {
            s[i] = hexDigits.substr(Math.floor(Math.random() * 0x10), 1);
        }
        s[14] = "4";  // bits 12-15 of the time_hi_and_version field to 0010
        s[19] = hexDigits.substr((s[19] & 0x3) | 0x8, 1);  // bits 6-7 of the clock_seq_hi_and_reserved to 01
        s[8] = s[13] = s[18] = s[23] = "-";

        var uuid = s.join("");
        return uuid;
    }

    var _sendData = function() {
        var self = this;

        var payload = self.analyticsData.payload || "";

        if (typeof(payload) === "object") {
            payload = JSON.stringify(payload);
        }

        self.analyticsData.payload = payload;

        var urlEncodedData = JSON.stringify(self.analyticsData);

        var http = new XMLHttpRequest();
        http.open("POST", self.url, true);
        http.setRequestHeader("Content-type", "application/json");
        http.send(urlEncodedData);
    };

    OTKAnalyticsComponent.prototype =  {
        constructor: OTKAnalytics,

        logEvent: function(data) { 
            this.analyticsData.action = data.action;
            this.analyticsData.variation = data.variation;

            //check values
            _checkData.call(this);

            //send data to analytics server
            _sendData.call(this);
        }
    }
    return OTKAnalyticsComponent;
})();
